addons/payment_razorpay/models/payment_provider.py

```diff
@@ -129,3 +186,283 @@ def _get_validation_amount(self):
+        :rtype: str
+        """
+        self.ensure_one()
+        OAUTH_LIVE_URL = self.env['ir.config_parameter'].get_param(
+            'OAUTH_LIVE_URL'
+        )
+        OAUTH__TEST_URL = self.env['ir.config_parameter'].get_param(
+            'OAUTH_TEST_URL'
+        )
+        if self.state == 'test':
+            return OAUTH__TEST_URL
+        else:
+            return OAUTH_LIVE_URL
+
+    # === ONBOARDING APIS ===#
+
+    def action_razorpay_connect_account(self):
+        """ Create a Razorpay Connect account and redirect the user.
+        If the provider is already enabled, close the current window. Otherwise
+        generate a Razorpay Connect onboarding link and redirect the user to it.
+        If the link generation fails, redirect the user to the provider form.
+
+        :return: URL for the next step.
+        :rtype: str
+        """
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_url',
+            'url': self._get_razorpay_authorize_url(),
+            'target': 'self',
+        }
+
+    def _razorpay_generate_state(self):
+        """ Generate a random 80-character string for use as a secure state parameter.
+
+        :return: Random string.
+        :rtype: str
+        """
+        return ''.join(secrets.choice(string.ascii_letters + string.digits) for _ in range(80))
+
+    def _get_razorpay_authorize_url(self):
+        """cGenerate the authorization URL for Razorpay.
+        This URL starts the authorization process with Razorpay and includes parameters such as
+        state, scope, redirect_uri, request_url, and uuid.
+
+        :return: Authorization URL.
+        :rtype: str
+        """
+        self.razorpay_authorization_state = self._razorpay_generate_state()
+        IrConfigParameter_sudo = self.env['ir.config_parameter'].sudo()
+        dbuuid = IrConfigParameter_sudo.get_param('database.uuid')
+        oauth_url = self._get_razorpay_oauth_url()
+        base_url = self.get_base_url()
+        return url_join(oauth_url, 'api/razorpay/1/authorize?%s' % url_encode({
+            'dbuuid': dbuuid,
+            'state': self.razorpay_authorization_state,
+            'redirect_url': base_url + '/payment/razorpay/oauth/callback',
+        }))
+
+    def cron_razorpay_refresh_token(self):
+        """ Cron job to refresh Razorpay access tokens.
+        This job updates tokens for records where the token will expire within the next day. It
+        searches for records with the code 'razorpay' and a matching refresh token, and then
+        attempts to refresh the token.
+        """
+        razorpay_need_refresh = self.search([
+            ('code', '=', 'razorpay'),
+            ('razorpay_refresh_token', '=', self.razorpay_refresh_token),
+            ('razorpay_access_token_expiration', '<=', fields.Datetime.now() + timedelta(days=1))])
+        for razorpay in razorpay_need_refresh:
+            try:
+                razorpay._razorpay_refresh_token()
```

Comment:
You can use `with self.env.cr.savepoint`
So it'll automatically rollback invaild data updated in that method